File: macos/FukuraMac/SnippetMatcher.swift
import Foundation

final class SnippetMatcher {
    private var buffer = ""
    private var snippets: [Snippet] = []
    private var maxTriggerLength = 0

    func update(snippets: [Snippet]) {
        self.snippets = snippets.sorted { left, right in
            if left.trigger.count == right.trigger.count {
                return left.trigger < right.trigger
            }
            return left.trigger.count > right.trigger.count
        }
        maxTriggerLength = self.snippets.map { $0.trigger.count }.max() ?? 0
        buffer = ""
    }

    func push(_ text: String) -> Snippet? {
        buffer += text
        if buffer.count > maxTriggerLength {
            buffer = String(buffer.suffix(maxTriggerLength))
        }

        if let match = snippets.first(where: { buffer.hasSuffix($0.trigger) }) {
            buffer = ""
            return match
        }
        return nil
    }

    func reset() {
        buffer = ""
    }
}

File: windows/SnippetMatcher.cs
namespace FukuraWindows;

internal sealed class SnippetMatcher
{
    private IReadOnlyList<Snippet> _snippets = [];
    private string _buffer = "";
    private int _maxTriggerLength;

    public void Update(IReadOnlyList<Snippet> snippets)
    {
        _snippets = snippets.OrderByDescending(item => item.Trigger.Length).ThenBy(item => item.Trigger).ToList();
        _maxTriggerLength = _snippets.Count == 0 ? 0 : _snippets.Max(item => item.Trigger.Length);
        Reset();
    }

    public Snippet? Push(string text)
    {
        if (_maxTriggerLength == 0) return null;
        _buffer += text;
        if (_buffer.Length > _maxTriggerLength) _buffer = _buffer[^_maxTriggerLength..];
        var match = _snippets.FirstOrDefault(item => _buffer.EndsWith(item.Trigger, StringComparison.Ordinal));
        if (match is not null) Reset();
        return match;
    }

    public void Reset() => _buffer = "";
}
